### ISSUE TITLE
OctoPrint API keys and CORS - tips for new users

### ISSUE THREAD

**Issue by davidzwa** [OctoFarm/OctoFarm#302]
To all OctoFarm users having difficulty getting the OctoPrint API key to work, @NotExpectedYet and me want to share a bit of information and a plan we have in the near future.

# API keys
There are 3 different types of API keys in OctoPrint. 
- User API Key (useful). The user API key works only if the user has the groups **Admins** and **Operator**. This is the first user you make, so make sure to add the roles yourself for it to work for another user (if you must).
- Application Key (useful). The same as a User key but targeted for specific applications: like OctoFarm. 
Make sure your user has the groups **ADMINS** AND **OPERATOR** for your user - this is default for only the 1st user!
- Global API Key. We've had trouble using the Global API Key and it's **will not fully work + should not be used => dont use it**.

Quoting OctoPrint's Application Key page:
> The global API key should best not be used anymore. It's one single key that gives full admin access to your whole OctoPrint instance. **Instead of using it you should create individual Application Keys for your third party clients.** That way they get permissions matching the user account used for key creation and you can also revoke access to one app without having to change the keys for all other apps. It's also recommended to create a user account without admin access and use that for third party clients where possible.

TL;DR create an Application Key for the first OctoPrint account you made ❤️ 
This the way to get there:
- Settings wrench icon 
![image](https://user-images.githubusercontent.com/6005355/108238363-2effcf00-7149-11eb-9bed-213cd2f29a52.png)
- Application Key (side navigation within OctoPrint Settings)
![image](https://user-images.githubusercontent.com/6005355/108238623-69696c00-7149-11eb-9b44-21b1362c1d23.png)
- Copy the generated key in the `Registered application keys` field which will appear after you click `Generate`
![image](https://user-images.githubusercontent.com/6005355/108240349-109ad300-714b-11eb-819b-7737fefeeff2.png)

- (Optional) you can do this also in the User Settings, where the same Application Key are managed but only for the user itself, as well as also a user-specific API Key can be generated. Pretty confusing, yeah. Just use Application Key.

## How to check the user's  groups
- (Optional) make sure the user has Operator and Admins groups (which is default for the first user). 
> OctoPrint Settings (wrench icon) > Access Control > Edit the user you want (crayon icon)

![image](https://user-images.githubusercontent.com/6005355/108240729-67a0a800-714b-11eb-9fa0-f551da95e749.png)

# CORS setting
What happens If CORS is disabled?
> The OctoFarm website (a.k.a. the client) cannot call OctoPrint. The OctoFarm server can still do it, but that's not how its currently all setup.

The CORS setting being checked off results in the website of OctoFarm not being able to OctoPrint directly. Thus stuff wil fail.
TL;DR enable CORS please ❤️ 

# What our plans are
Right now if you enter a Global Key, we can use a small hack to check if it is indeed the global key and try to throw a warning. This is not something permanent and you should follow the instructions above.

About CORS, we will try and enable it for you, in a future release, given a correct API Key is provided to reach the OctoPrint client. This is not a guarantee on our side however.

Happy printing!
The devs

**Comment by NotExpectedYet** [comment]
Legend thanks for that man!

**Comment by NotExpectedYet** [comment]
No stale

**Comment by davidzwa** [comment]
Ill play bot haha

**Comment by lsls421** [comment]
Hi! Thanks for the effort of creating this guide, it is much appreciated. 

I would like to ask a short question: I have two admin user accounts in OctoPrint, and I actually want to use the second user's API for OctoFarm. Is it possible to tell OctoFarm to choose between multiple admin users instead of automatically choosing the first one?

**Comment by NotExpectedYet** [comment]
> Hi! Thanks for the effort of creating this guide, it is much appreciated.
> 
> I would like to ask a short question: I have two admin user accounts in OctoPrint, and I actually want to use the second user's API for OctoFarm. Is it possible to tell OctoFarm to choose between multiple admin users instead of automatically choosing the first one?

Not currently no sadly. OctoFarm was written to save a bit of time in that regards for larger farms. It's worth adding a discussion for a feature request and we will look into adding it though. :)

**Comment by lsls421** [comment]
> > Hi! Thanks for the effort of creating this guide, it is much appreciated.
> > I would like to ask a short question: I have two admin user accounts in OctoPrint, and I actually want to use the second user's API for OctoFarm. Is it possible to tell OctoFarm to choose between multiple admin users instead of automatically choosing the first one?
> 
> Not currently no sadly. OctoFarm was written to save a bit of time in that regards for larger farms. It's worth adding a discussion for a feature request and we will look into adding it though. :)

Thanks for the quick feedback. I can understand the reason for this. Feature request added in the discussion. :D

**Comment by davidzwa** [comment]
@github-actions[bot] not stale my friend
